### ISSUE TITLE
Each item in the matrix should get its own cache

### ISSUE THREAD

**Issue by wereHamster** [travis-ci/travis-ci#4657]
The [.travis.yml template for Haskell](https://github.com/hvr/multi-ghc-travis) currently uses `language: c` for the only reason so that travis creates separate caches for each item in the build matrix (different GHC compilers). I noticed that there is `language: generic`, which would be a perfect fit for Haskell, but it uses a single cache. Which made me thinking..

The items in the build matrix are built in parallel, when they use the same cache, they overwrite each other! Even if the toolchain was capable of using a single cache for multiple compiler versions (cabal/GHC certainly is), the current setup won't work. Now, *is* there a reason to use a single cache, other than for storage space optimization?

I see two options:

 1. If there is a way to uniquely identify each item in the matrix (e.g. by hashing the whole item), use that as the cache slug.
 2. Give users an option to set the cache slug for each item in the matrix, but make it very clear in the documentation that there is a race condition when storing the cache. Ideally that should work regardless of the language setting.

I prefer the first one because it requires no configuration on the users side. But it may explode the used storage space in your S3 bucket.

**Comment by BanzaiMan** [comment]
Duplicate of #4393.
